File: src/main/java/org/veupathdb/lib/blast/field/HSPSorting.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.JSONConstructor;
import org.veupathdb.lib.blast.util.JSONValue;

public enum HSPSorting implements JSONValue
{
  ByHSPExpectValue,
  ByHSPScore,
  ByHSPQueryStart,
  ByHSPPercentIdentity,
  ByHSPSubjectStart;

  public int getValue() {
    return this.ordinal();
  }

  public static HSPSorting fromIntValue(int value) {
    return values()[value];
  }

  @JsonCreator
  public static HSPSorting fromJSON(JsonNode js) {
    return fromIntValue(js.intValue());
  }

  @Override
  @JsonValue
  public JsonNode toJSON() {
    return JSONConstructor.newInt(this.ordinal());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/HSPSorting.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.JSONConstructor
import com.fasterxml.jackson.annotation.JsonCreator

enum class HSPSorting : JSONValue {
  ByHSPExpectValue, ByHSPScore, ByHSPQueryStart, ByHSPPercentIdentity, ByHSPSubjectStart;

  val value =  ordinal

  @JsonValue
  override fun toJSON(): JsonNode {
    return JSONConstructor.newInt(ordinal)
  }

  companion object {
    @JvmStatic
    fun fromIntValue(value: Int): HSPSorting {
      return values()[value]
    }

    @JsonCreator
    @JvmStatic
    fun fromJSON(js: JsonNode): HSPSorting {
      return fromIntValue(js.intValue())
    }
  }
}